### PR TITLE
feat: Add Dockerfile, tests, and CI

### PR DESCRIPTION
**Pull request by matthewfeickert**
Add a Dockerfile, tests for the built image, and CI to build and publish the Docker image.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add Dockerfile
   - Based on Conda-forge root_base package, as have source build errors
   - c.f. https://root-forum.cern.ch/t/segfault-from-numpy-import-after-root-in-pyroot-in-debian-docker-image/38265
* Add tests to ensure RooFit works, and that the import order of ROOT and NumPy is safe
* Add CI using GitHub Actions to test the image build and to publish to Docker Hub
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
For reasons I do not understand at the moment, things are not import order safe.

This runs
```
docker run \
	--rm \
	pyhf/pyhf-validation-root-base:latest \
	-c "python -m pip install numpy; python -c 'import numpy as np; import ROOT; print(np.arange(10))'"
```

but this segfault

```
docker run \
	--rm \
	pyhf/pyhf-validation-root-base:latest \
	-c "python -m pip install numpy; python -c 'import ROOT; import numpy as np; print(np.arange(10))'"
```

PyROOT must be doing something strange. :?

**Comment by matthewfeickert** [comment]
I've added an question on ROOT Forums now: [Segfault from NumPy import after ROOT in PyROOT in Debian Docker image](https://root-forum.cern.ch/t/segfault-from-numpy-import-after-root-in-pyroot-in-debian-docker-image/38265)

File: Dockerfile.bak [hunk 1]
```diff
@@ -0,0 +1,79 @@
+ARG BASE_IMAGE=python:3.7-slim
```
Comment (kratsg) [inline Dockerfile.bak]: delete this file

Comment (matthewfeickert) [reply Dockerfile.bak]: Yes of course. This is why the PR is still in draft mode.

**Comment by matthewfeickert** [comment]
If I use a Conda-forge `root_base` based build (HT to @henryiii and @chrisburr) then I'm able to get ROOT and NumPy to play nice together in a Python 3 environment. This results in a larger Docker image then doing a pure source build, but it is _much_ faster and it currently works. So for those reasons I will take a big something over a smaller nothing.

**Review by kratsg — APPROVED** [review]

**Comment by henryiii** [comment]
How much larger, out of curiosity? Haven’t seen a `root_base` based number yet. :)

**Comment by matthewfeickert** [comment]
> How much larger, out of curiosity? Haven’t seen a `root_base` based number yet. :)

These are the uncompressed image sizes (`root_base-build-latest` is what is [currently on Docker Hub as `latest`](https://hub.docker.com/layers/pyhf/pyhf-validation-root-base/latest/images/sha256-2bc08db058b115276b4a2f62844584ead5cbf51db5a79bf2f7c89ff70e4cadda?context=explore))

```
$ docker images pyhf/pyhf-validation-root-base
REPOSITORY                       TAG                     IMAGE ID            CREATED             SIZE
pyhf/pyhf-validation-root-base   root_base-build-latest  bd8202622fe0        10 hours ago        1.46GB
pyhf/pyhf-validation-root-base   root-build-latest       9f77ef1eca06        30 hours ago        1.56GB
pyhf/pyhf-validation-root-base   src-build-latest        7c90a40c9ec4        8 days ago          842MB
```

**Review by lukasheinrich — APPROVED** [review]

**Pull request merged**